### PR TITLE
Feature - conform templates script

### PR DESCRIPTION
**Pull request by webimpress**
It should be called from main repository package:
```console
$ zf-maintainer rebase-doc-templates <path-to-package>
```
It takes package name from composer "name" key.

- renames `doc` -> `docs` if exists (otherwise creates `docs` directory)
- puts all support files and templates into `docs` directory
- removes CONTRIBUTING.md and CONDUCT.md from main directory
- updates LICENSE.md
- updates mkdocs.yml if exists (doc -> docs and copyright year)
- updates coveralls.yml
- updates .gitattributes
- updates .gitignore
- adds .travis.yml only file in repo does not exist
- updated composer.json
  - checks repository description and update it to match with description in
    https://docs.zendframework.com/zf-mkdoc-theme/scripts/zf-component-list.json
  - fixes sections order
  - removes default type if it's library
  - updates scripts
  - updates license
  - updates support
  - updates config
  - removes "minimum-stability" and "prefer-stable"
  - checks keywords
  - checks branch-alias (dev-master and dev-develop)
- updates badges in README.md and prints error when badge is not found
- replaces links to the documentation (docs.zendframework.com) in README.md and `docs/books` md files

I've used this script to prepare PR in all `zend-expressive-*` packages and tested also in couple others.
Definitely we can improve it more, but as starter I think it's good enough ;-)

### CODE REVIEW

**Comment by weierophinney** [comment]
And... just for the record... I have created [weierophinney/zf-component-template](https://github.com/weierophinney/zf-component-template), which can be used when creating a new component. It is in-sync currently with this repository, and includes as self-removing `setup.php` script that does these same operations. It also provides a `ConfigProvider` class, and a test for that class, as a baseline for the new component.

This stuff _could_ be moved into this repository as well. The main difference is that creation wouldn't be clone + setup, but instead just invoking the script.

File: bin/conform-templates.php [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    }
+}
+file_put_contents('.gitattributes', $content);
+
+// .gitignore
+$content = file_get_contents(__DIR__ . '/../template/.gitignore');
+if (! $hasDocs) {
+    $content = preg_replace('/^docs\/html\/\s+/m', '', $content);
+    $content = preg_replace('/^zf-mkdoc-theme\/\s+/m', '', $content);
+    $content = preg_replace('/^zf-mkdoc-theme\.tgz\s+/m', '', $content);
```
Comment (weierophinney) [inline bin/conform-templates.php]: Do the same as I suggested above: use an alternate regexp delimiter, and use the `s` modifier to capture the trailing EOL with the `\s*` set.

File: bin/conform-templates.php [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+$templateContent = json_decode(file_get_contents(__DIR__ . '/../template/composer.json'), true);
+
+$sectionOrder = [];
+foreach ($templateContent as $section => $value) {
+    $sectionOrder[] = $section;
+}
+
+$content = $composer;
+if (isset($content['type']) && $content['type'] === 'library') {
+    unset($content['type']);
```
Comment (weierophinney) [inline bin/conform-templates.php]: Why are you unsetting this? Shouldn't this stay?

File: bin/conform-templates.php [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    if (! $hasZf) {
+        array_unshift($content['keywords'], 'zf');
+    }
+}
+
+if (file_exists('.docheader')) {
+    array_unshift($content['scripts']['check'], '@license-check');
+    $content['scripts']['license-check'] = 'docheader check src/ test/';
+}
```
Comment (weierophinney) [inline bin/conform-templates.php]: These two operations should only be done _if those elements are not already defined_.

File: bin/conform-templates.php [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    return 1;
+});
+
+file_put_contents(
+    'composer.json',
+    json_encode(
+            $content,
+            JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES
+    ) . PHP_EOL
+);
```
Comment (weierophinney) [inline bin/conform-templates.php]: After this is done, we need to run a `composer update --lock`; otherwise, composer will complain about the lockfile being out-of-sync with the `composer.json` file. (I've experienced this a number of times...)

File: bin/conform-templates.php [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    $content = preg_replace('/site_dir:.*/', 'site_dir: docs/html', $content);
+
+    if (preg_match('/Copyright \(c\) (\d{4})/', $content, $m)) {
+        $year = $m[1];
+        $copyrightYear = $year < $currentYear ? $year . '-' . $currentYear : $currentYear;
+        $content = preg_replace(
+            '/Copyright \(c\) \d{4}(-\d{4}) /',
+            'Copyright (c) ' . $copyrightYear . ' ',
+            $content
+        );
```
Comment (weierophinney) [inline bin/conform-templates.php]: I'd argue that we should reuse the `$yearReplacement` value I proposed earlier. The above then becomes the following:

```php
if (preg_match('/Copyright \(c\) (\d{4}(\s*-\s*\d{4})?)/s', $content, $matches)) {
    $content = str_replace($matches[1], $yearReplacement, $content);
}
```

File: bin/conform-templates.php [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        file_put_contents('mkdocs.yml', $content);
```
Comment (weierophinney) [inline bin/conform-templates.php]: We want to rewrite the contents regardless of whether or not we were able to update the copyright date; this should be moved outside the conditional.

File: bin/conform-templates.php [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+$content = str_replace(
+    '{year}',
+    $year < $currentYear ? $year . '-' . $currentYear : $currentYear,
+    $content
+);
+
+file_put_contents('LICENSE.md', $content);
+
+// .coveralls.yml
+file_put_contents('.coveralls.yml', file_get_contents(__DIR__ . '/../template/.coveralls.yml'));
```
[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: Never mind; I see you added it in this patch as well.

File: template/.gitattributes [hunk 8]
```diff
@@ -1,10 +1,10 @@
+/docs export-ignore
+/test export-ignore
```
Comment (weierophinney) [inline template/.gitattributes]: Why did you change this? The values were sorted via `natsort` previously...

File: template/composer.json [hunk 9]
```diff
@@ -23,6 +23,15 @@
...
     },
+    "provide": {
+        "psr/package": "<version>"
+    },
+    "conflict": {
+        "conflicted/package": "<version>"
+    },
+    "suggest": {
+        "suggested/package": "<suggested version> <description>"
+    },
```
Comment (weierophinney) [inline template/composer.json]: I don't think these should be in the template, as they are (1) optional, and (2) will vary greatly for each package. They can be added if/when they are needed.

File: template/composer.json [hunk 10]
```diff
@@ -44,6 +53,9 @@
             "config-provider": "{namespace}\\ConfigProvider"
         }
     },
+    "bin": [
+        "<binary/file.php>"
+    ],
```
Comment (weierophinney) [inline template/composer.json]: Same applies here.

File: bin/conform-templates.php [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+$hasDocs = file_exists('mkdocs.yml');
+
+// .gitattributes
+$content = file_get_contents(__DIR__ . '/../template/.gitattributes');
+if (! $hasDocs) {
+    $content = preg_replace('/\/mkdocs\.yml export-ignore\s+/', '', $content);
+}
+if (file_exists('phpbench.json')) {
+    if (file_exists('benchmark') && is_dir('benchmark')) {
+        $content = '/benchmark export-ignore' . PHP_EOL . $content;
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: I'd like to keep alphabetical order in `.gitattributes` - first directories, then files.
It's a bit harder on adding new entry in this script but I think it's worth to do that :)

File: bin/conform-templates.php [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    if (file_exists('benchmarks') && is_dir('benchmarks')) {
+        $content = '/benchmarks export-ignore' . PHP_EOL . $content;
+    }
+    $content = preg_replace(
+        '/\/phpcs\.xml export-ignore/',
+        '/phpbench.json export-ignore' . PHP_EOL . '/phpcs.xml export-ignore',
+        $content
+    );
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: It will be then alphabetically but files will be mixed with dirs. Usually files are sorted - first directories, then files (this is what I can see in project struct in PHPStorm.

I'll check what I can do with it, I agree that it's not so clear right now.

File: bin/conform-templates.php (hunk 2 above)
Comment (webimpress) [reply bin/conform-templates.php]: I'm unsetting only in case of "type=library" as it is default type.
See https://getcomposer.org/doc/04-schema.md#type
And https://github.com/zendframework/zend-stdlib/pull/85#pullrequestreview-71986270

File: bin/conform-templates.php (hunk 3 above)
Comment (webimpress) [reply bin/conform-templates.php]: I override all scripts before to match these from templates - see line 179, and these are not defined. Anyway, I think we are going to remove it soon - and use licence check from zend-coding-standards library.

File: bin/conform-templates.php (hunk 4 above)
Comment (webimpress) [reply bin/conform-templates.php]: Actually not, because we are not changing any requirements, I've checked it with 1.5.2 version.

File: template/.gitattributes (hunk 8 above)
Comment (webimpress) [reply template/.gitattributes]: Alphabetical order, directories first, the same we have in .gitignore
Also the same order I can see in PHPStorm project view.

File: template/composer.json (hunk 9 above)
Comment (webimpress) [reply template/composer.json]: I've added it because I'm using that template in the script to sort sections always in the same order. We can set sections order in the script and then we don't need to add it here. What do you think?

File: template/composer.json (hunk 10 above)
Comment (webimpress) [reply template/composer.json]: See comment above.

File: bin/conform-templates.php [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+        fwrite(STDERR, 'Cannot match year in current LICENSE.md file. Using current year only.' . PHP_EOL);
+    }
+}
+
+$content = file_get_contents(__DIR__ . '/../template/LICENSE.md');
+$content = str_replace(
+    '{year}',
+    $year < $currentYear ? $year . '-' . $currentYear : $currentYear,
+    $content
+);
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: In your code this part of regular expression is not used at all: `(?:\s*-\s*\d{4})?`.
As you noted we need to find only "startYear" and this is exactly what my regular expression is looking for. I'll rename just variables in my solution and it should be clearer then.

[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: My main point was making it clear that either a single year or range of years could be replaced. But you're right; the only bit _the code_ is interested in is the start year, so you could eliminate that part of the regexp.

File: bin/conform-templates.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: See my solution regarding the `.gitignore` file; make all changes, and then do a `natsort` of the entries when done.

One thing you could do is explode the entries initially, and then do one or more `$content[] = ...` lines; then, when all items are aggregated, you would do `file_put_contents($filename, implode("\n", $content)` operation.

Either way, it simplifies the code dramatically, and doesn't make assumptions about the existing order of entries.

File: bin/conform-templates.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: Directories will have a `/` following them generally, which should make sorting work.

File: bin/conform-templates.php (hunk 2 above)
Comment (weierophinney) [reply bin/conform-templates.php]: Okay; wasn't sure about that as I recently saw PRs that added the `"type": "library"` entry.

File: bin/conform-templates.php (hunk 3 above)
Comment (weierophinney) [reply bin/conform-templates.php]: Exactly. :)

File: bin/conform-templates.php (hunk 4 above)
Comment (weierophinney) [reply bin/conform-templates.php]: That must be a recent fix in the 1.5 series, then. If that's the case, no change needed.

File: bin/conform-templates.php [hunk 14]
```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env php
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Done

File: template/.gitattributes (hunk 8 above)
Comment (weierophinney) [reply template/.gitattributes]: That would be more clear if directories had a `/` suffix; without it, on unix systems at least, files do not need suffixes, which means it's hard to determine if the entry is a directory or a file.

Alphabetical order is the most important, as far as I'm concerned; it makes finding the entry in the list easier.

File: bin/conform-templates.php [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+    exit(1);
+}
+$org = $m[1];
+$repo = $m[2];
+
+// if there is no docs directory create one
+if (! is_dir('docs')) {
+    // if there is doc directory rename it to docs
+    if (is_dir('doc')) {
+        rename('doc', 'docs');
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Hm... I think I wrote my comment here, but I can't see it now. I've tested it on couple repositories and it seems that GIT recognizes rename automatically. I can change it to system command if you prefer.

File: bin/conform-templates.php [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+    'CONTRIBUTING.md',
+    'ISSUE_TEMPLATE.md',
+    'PULL_REQUEST_TEMPLATE.md',
+    'SUPPORT.md',
+];
+
+$replace = [
+    '{org}' => $org,
+    '{repo}' => $repo,
+    '{category}' => strpos($repo, 'zend-expressive') === 0 ? 'expressive' : 'components',
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Done

File: template/composer.json (hunk 9 above)
Comment (weierophinney) [reply template/composer.json]: Do it in the script, and not the `composer.json`, please; that way, we don't need to manually update the `composer.json` immediately for new components to make it work. WIth these entries, you'll immediately get failures if you run `composer validate` otherwise, even after making substitutions of `{var}` entries.

File: bin/conform-templates.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Updated in 21bedf0

File: bin/conform-templates.php [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+// .coveralls.yml
+file_put_contents('.coveralls.yml', file_get_contents(__DIR__ . '/../template/.coveralls.yml'));
+
+// check if repository has documentation
+$hasDocs = file_exists('mkdocs.yml');
+
+// .gitattributes
+$content = file_get_contents(__DIR__ . '/../template/.gitattributes');
+if (! $hasDocs) {
+    $content = preg_replace('/\/mkdocs\.yml export-ignore\s+/', '', $content);
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Changed in 3853ff8

File: bin/conform-templates.php [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+}
+if (file_exists('phpbench.json')) {
+    if (file_exists('benchmark') && is_dir('benchmark')) {
```
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: Good point, done.

File: bin/conform-templates.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (webimpress) [reply bin/conform-templates.php]: See 3853ff8

File: bin/conform-templates.php (hunk 1 above)
Comment (webimpress) [reply bin/conform-templates.php]: Changed to `array_diff`

File: bin/conform-templates.php (hunk 5 above)
Comment (webimpress) [reply bin/conform-templates.php]: Changed - see 21bedf0

File: bin/conform-templates.php (hunk 6 above)
Comment (webimpress) [reply bin/conform-templates.php]: Good point, fixed in af5c781

File: bin/conform-templates.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: I was thinking more in terms of the steps needed to run this script.

They become:

- run script
- remove old entries (`git rm -rf doc`)
- add new entries (`git add doc`)

While git will detect these correctly as renames, it's three operations, instead of one, which is my main concern.

File: bin/conform-templates.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (weierophinney) [reply bin/conform-templates.php]: I've added a comment to that commit - overall, definitely an improvement!

**Comment by weierophinney** [comment]
> According to wikipedia \n\r is possible somewhere:

Yes, RiscOS and Acorn BBS... which PHP doesn't support anyways. I mean, I _suppose_ you could edit on these platforms and deploy to another platform, but it's unlikely.

I've been working with line-endings in PHP OSS projects for over a dozen years, and have literally never had anybody ask to support LF+CR. :smile:

**Comment by weierophinney** [comment]
With #28 merged, do you want me to merge this, or wait for you to refactor it to symfony/console?

**Comment by webimpress** [comment]
@weierophinney 
> With #28 merged, do you want me to merge this, or wait for you to refactor it to symfony/console?

I'm refactoring it now to use symfony/console. Will be ready shortly.

**Comment by webimpress** [comment]
@weierophinney It's now updated

**Comment by weierophinney** [comment]
Thanks, @webimpress!
